File: python/sorting_algorithms/quicksort.py
def part(arr, low, high):
    pivot = arr[high]
    low_cache = low
    high_cache = high

    while low < high:
        while low < high_cache and arr[low] < pivot:
            low += 1
        
        while high >= low_cache and arr[high] >= pivot:
            high -= 1
        
        if low < high:
            arr[low], arr[high] = arr[high], arr[low]
    
    if arr[low] > pivot:
        arr[low], arr[high_cache] = arr[high_cache], arr[low]
    
    return low

def quicksort(arr, low=0, high=None):
    if high is None:
        high = len(arr) - 1
    
    if low < high:
        pivot = part(arr, low, high)
        quicksort(arr, pivot+1, high)
        quicksort(arr, low, pivot-1)
    
    return arr

File: c++/sortingAlgorithms/quicksort.cpp
#include "utils.cpp"


int partition(int arr[], int low, int high) {
    int pivot = arr[high];
    int l = (low - 1);

    for (int pos = low; pos < high; pos++) {
        if (arr[pos] <= pivot) {
            l++;
            swap(&arr[l], &arr[pos]);
        }
    }
    
    swap(&arr[l+1], &arr[high]);

    return (l + 1);
}

void quicksort(int arr[], int low, int high) {
    if (low < high) {
        int pivot = partition(arr, low, high);

        quicksort(arr, pivot+1, high);
        quicksort(arr, low, high-1);
    }
}




int main()
{
    print_array(arr, length);

    quicksort(arr, 0, length-1);
    print_array(arr, length);

    cout << (equal_arrays(arr, sol, length) ? "true" : "false");

    return 0;
}